soh/src/overlays/misc/ovl_kaleido_scope/z_kaleido_progress.c

```diff
@@ -0,0 +1,456 @@
+        X_Offset = X_Offset_GS[gSaveContext.language][ID];
+        Y_Offset = 1;
+    }
+    if (Icon_Type == TYPE_SMALL_KEYS) {
+        X_Offset = -4;
+    }
+    if (Icon_Type == TYPE_MAP) {
+        X_Offset = 15;
+    }
+    if (Icon_Type == TYPE_COMPASS) {
+        X_Offset = 30;
+    }
+    if (Icon_Type == TYPE_BOSS_KEY) {
+        X_Offset = 45;
+    }
+    if (Icon_Type == TYPE_TOKEN) {
+        X_Offset = 60;
+    }
+    s16 PosX = MENU_LEFT_INNER + (MENU_CELLS_SPACE_X * Col)+X_Offset; //General Pos
+    s16 PosY = MENU_TOP_INNER + (MENU_CELLS_SPACE_Y * Line)+Y_Offset; //General Pos
+    int height = sprite.height * sprite.scale_h; //Adjust Height with scale
+    int width = sprite.width * sprite.scale_w; //Adjust Width with scale
+    int height_factor = (1 << 10) * sprite.height / height;
+    int width_factor = (1 << 10) * sprite.width / width;
+
+    //We open our Disp that will allow graphic draw
+    OPEN_DISPS(globalCtx->state.gfxCtx, "z_kaleido_progress.c", 100);
+    gDPPipeSync(POLY_KAL_DISP++);
+    gDPSetCombineMode(POLY_KAL_DISP++, G_CC_MODULATEIA_PRIM, G_CC_MODULATEIA_PRIM);
+
+    //We set color to match the one we set in the Sprite.
+    if(greyscale) {
+        gsDPSetGrayscaleColor(POLY_KAL_DISP++, 109, 109, 109, 220);
+        gsSPGrayscale(POLY_KAL_DISP++, true);
+    } else {
+        gDPSetPrimColor(POLY_KAL_DISP++, 0, 0, 255, 255, 255, 255);
+        gsSPGrayscale(POLY_KAL_DISP++, false);
+    }
+
+    //Loading texture with proper settings (settings like size if it clamp mirror etc.)
+    gDPLoadTextureBlock(POLY_KAL_DISP++, sprite.tex, G_IM_FMT_RGBA, G_IM_SIZ_32b, sprite.width, sprite.height, 0, G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMASK, G_TX_NOMASK, G_TX_NOLOD, G_TX_NOLOD);
+
+    //Making a "room" for our loaded graphic, it's like making a size where our graphic will draw, it has a top left corner at one postion and bottom right corner at another and make a rectangle from it, and we draw in that rectangle
+    gSPTextureRectangle(POLY_KAL_DISP++, PosX << 2, PosY << 2, (PosX + width) << 2, (PosY + height) << 2, G_TX_RENDERTILE, 0, 0, width_factor, height_factor);
+
+    //We close the DISP has there no more GFX to be draw there.
+    gsSPGrayscale(POLY_KAL_DISP++, false);
+    gDPPipeSync(POLY_KAL_DISP++);
+    CLOSE_DISPS(globalCtx->state.gfxCtx, "z_kaleido_progress.c", 400);
+}
+
+void KeleidoScope_DungeonContainerDraw(GlobalContext* globalCtx) {
+                                        //Right now the structure has a count for the maximumm small key possible but do not use it maybe 
+                                        //I could make a system to keep track of how many keys the players had and have and 
+                                        //once he/she reach the max amount it change the digit to green?
+    Kaleido_dgn_container DungeonsSettings[] = {{ DekuTitle_Kspr, 0, false, false, false, true, true, false, true, KokiriT_Kspr,false,0,0,0 },
```

Comment:
This should come outside of the function. Or if you want to keep it in the function make it `static` so its not copied on each call.

Reply:
should be correct now ~
Btw you made me learn something today thank you, I had no idea a structure would be called each time but that logic.
It made me search how to redo it with pointer and it work just as fine and I hope this time it is correct :)